crates/cargo-gpu/src/main.rs

```diff
@@ -49,681 +49,50 @@
-    ///
-    /// Pretty much runs:
-    ///
-    /// * rustup toolchain add nightly-2024-04-24
-    /// * rustup component add --toolchain nightly-2024-04-24 rust-src rustc-dev llvm-tools
-    fn ensure_toolchain_and_components_exist(&self) {
-        // Check for the required toolchain
-        let output = std::process::Command::new("rustup")
-            .args(["toolchain", "list"])
-            .output()
-            .unwrap();
-        assert!(
-            output.status.success(),
-            "could not list installed toolchains"
-        );
-        let stdout = String::from_utf8_lossy(&output.stdout);
-        if stdout
-            .split_whitespace()
-            .any(|toolchain| toolchain.starts_with(&self.channel))
-        {
-            log::debug!("toolchain {} is already installed", self.channel);
-        } else {
-            let output = std::process::Command::new("rustup")
-                .args(["toolchain", "add"])
-                .arg(&self.channel)
-                .stdout(std::process::Stdio::inherit())
-                .stderr(std::process::Stdio::inherit())
-                .output()
-                .unwrap();
-            assert!(
-                output.status.success(),
-                "could not install required toolchain"
-            );
-        }
-
-        // Check for the required components
-        let output = std::process::Command::new("rustup")
-            .args(["component", "list", "--toolchain"])
-            .arg(&self.channel)
-            .output()
-            .unwrap();
-        assert!(
-            output.status.success(),
-            "could not list installed components"
-        );
-        let stdout = String::from_utf8_lossy(&output.stdout);
-        let required_components = ["rust-src", "rustc-dev", "llvm-tools"];
-        let installed_components = stdout.lines().collect::<Vec<_>>();
-        let all_components_installed = required_components.iter().all(|component| {
-            installed_components.iter().any(|installed_component| {
-                let is_component = installed_component.starts_with(component);
-                let is_installed = installed_component.ends_with("(installed)");
-                is_component && is_installed
-            })
-        });
-        if all_components_installed {
-            log::debug!("all required components are installed");
-        } else {
-            let output = std::process::Command::new("rustup")
-                .args(["component", "add", "--toolchain"])
-                .arg(&self.channel)
-                .args(["rust-src", "rustc-dev", "llvm-tools"])
-                .stdout(std::process::Stdio::inherit())
-                .stderr(std::process::Stdio::inherit())
-                .output()
-                .unwrap();
-            assert!(
-                output.status.success(),
-                "could not install required components"
-            );
-        }
-    }
-}
+mod builder;
```

Comment:
This is a bit of a nitpick, but I think we should rename this module to `build`, just because when I first saw the name of the file I thought it would contain a struct with a builder pattern. Is that reasonable?

Reply:
I totally agree! What happened is that I did originally call it `build` but `rust-analyzer` seemed to get confused about the `build.rs` file thinking it was that pre-compilation compilation convention.

But now that I try again, it works fine. So back to `build` it is 🎉